### ISSUE TITLE
Subsequent subtasks are run before previous ones finish

### ISSUE THREAD

**Issue by pianocomposer321** [pianocomposer321/yabs.nvim#12]
See https://github.com/pianocomposer321/yabs.nvim/issues/5#issuecomment-932214897:

> I have 3 tasks for cpp: build, run and clean and default_task is { "build", "run", "clean" }. However it looks like they run out of order sometimes. i.e clean then run then build

This happens because the tasks, which are asynchronous by design, don’t wait for the previous ones to finish before starting. This means that if earlier commands take any amount of time to complete, subsequent ones will start before they are finished executing.

This is a very difficult problem, because the solution is different depending on the output type. Most of the output types use luv under the hood, but `terminal` and `consolation` do not. For those using luv, I could probably figure something out that would work for all of them, but that would not solve the problem for terminal outputs. It is possible that I could hack something together by checking the pids of the child processes of the terminals, but this is not without issues of its own (mostly portability). Not to mention, even if the hack did work (which is a big if), it wouldn’t solve the problem for user-defined outputs.

I’m really not sure how to go about solving this one. Short of a massive redesign of pretty much the entire plugin, there’s not much we can do. Unless I think of something or someone else suggests something else that I haven’t thought of, there may not be much of a choice but to remove the default task being defined as multiple subtasks feature all together.

CC: @Shatur @p00f

**Comment by Shatur** [comment]
Unfortunately I am not quite familiar with luv. But doesn't plenary provide anything for running commands down the chain?

**Comment by pianocomposer321** [comment]
Possibly, but if so, they’re just abstractions on what’s already provided in luv. They still wouldn’t solve the problem for outputs that don’t use luv, nor for user-defined ones.

**Comment by p00f** [comment]
If you are willing to ignore consolation and terminal outputs and only support buffer, you can use the viml functions jobstart() and jobwait()

**Comment by p00f** [comment]
My bad,  I didn't read your comment fully

**Comment by p00f** [comment]
Maybe only allow quickfix and buffer if the task is a list?

**Comment by pianocomposer321** [comment]
> Maybe only allow quickfix and buffer if the task is a list?

I’m not sure I understand how that would solve the problem…

**Comment by pianocomposer321** [comment]
I think what I might do is drop support for the subtasks feature and instead add a third task type, `lua` or `function` or `lua_function` or something like that, that just runs the provided function as-is. Technically you can already do this by just having `command` return an empty string, but it’s not officially supported and I’m actually not 100% sure that it wouldn’t break anything.

Anyway, with this new task type, you could just run `require(“yabs”):run_task(taskname)` multiple times with your own logic for when to do so. That would shift the responsibility for doing that from the plugin to the user, which obviously has pros and cons, but a big pro is that I wouldn’t have to modify how the outputs fundamentally function. Most of the time, I think that running the program in the terminal would always be the last step in the chain, but if I was still planning to support chained tasks, I would have to come up with a solution for that anyway. If it’s up to the user, all I have to do is make sure they’re aware of the shortcomings and let them figure it out themselves.

I would probably have to expose a few more apis for checking the status of jobs, so there’s definitely more work that would need to be done, but I think it’s a more elegant solution, and at the very least a plausible one.

LMK what you think.

**Comment by Shatur** [comment]
Can we concatenate commands via `&&` for `terminal` and `consolidation`?

BTW, there is no chained command support in Asyntasks, but you can get the job status and there is an autocmd that can be used to run something in the chain. This is what I use in my neovim-cmake plugin.

**Comment by pianocomposer321** [comment]
> Can we concatenate commands via `&&` for `terminal` and `consolidation`?

Yes, whatever text you set for command is passed directly to the output function as-is. That means that for `terminal` and `consolation` outputs, the command is sent to the terminal, `&&` and all.

> BTW, there is no chained command support in Asyntasks, but you can get the job status and there is an autocmd that can be used to run something in the chain. This is what I use in my neovim-cmake plugin.

Cool! I’ll probably look there before making a final decision. Could you point me to either where it is implemented in asynctasks (or asyncrun) or where you use it in your plugin?

**Comment by Shatur** [comment]
> Yes, whatever text you set for command is passed directly to the output function as-is. That means that for terminal and consolation outputs, the command is sent to the terminal, && and all.

So we can use `&&` for `consolation` and `terminal`? But there is no solution for quickfix?

> Could you point me to either where it is implemented in asynctasks (or asyncrun) or where you use it in your plugin?

I'm not familiar with asynctasks code, but here is how I assign a command for autocmd: https://github.com/Shatur/neovim-cmake/blob/c2dbed01c7b27fd55105195cf6bab6761f543b2f/lua/cmake/utils.lua#L109

**Comment by pianocomposer321** [comment]
> So we can use `&&` for `consolation` and `terminal`? But there is no solution for quickfix?

Nope, as far as I know `&&` should work for quickfix et al as well. The problem is when you want to chain commands together that are being sent to different outputs. For example, if you want the build output to go to the quickfix list and then the run output to go to the terminal.
 
> I'm not familiar with asynctasks code, but here is how I assign a command for autocmd: https://github.com/Shatur/neovim-cmake/blob/c2dbed01c7b27fd55105195cf6bab6761f543b2f/lua/cmake/utils.lua#L109

Thanks. I actually found where it is in the asynrun code, and I’m taking a look now. I’ll get back with you.

**Comment by Shatur** [comment]
> The problem is when you want to chain commands together that are being sent to different outputs

Completely forgot about it, you are right :(

**Comment by pianocomposer321** [comment]
@Shatur question:
Does neovim-cmake work if you set the output type for asyncrun to terminal?

**Comment by pianocomposer321** [comment]
Ok, so I’m trying to figure out how asyncrun uses the autocommands, but they don’t seem to be working at all for me…

Here’s the code I’m using to test it, in a file called `lua/utils.lua`:

```lua
local M = {}

-- ...

function M.asyncrun_callback(function_string)
    vim.api.nvim_command(‘autocmd User AsyncRunStop ++once if g:asyncrun_status ==? “success” | echo “executing callback” | call luaeval(“‘ .. function_string .. ‘“) | endif’)
end

function M.build()
    local command = “echo building…”
    vim.fn[‘asyncrun#run’](‘’, {mode = 1}, command}
end

function M.run()
    local command = “echo running…”
    vim.fn[‘asyncrun#run’](‘’, {mode = 1}, command}
end

function M.build_and_run()
    M.asyncrun_callback(“require(‘utils’).run()”)
    M.build()
end

-- ...

return M
```

And I’m running it with `:lua require(“utils”).build_and_run()`. I see “building…”, but not “executing callback” or “running…”.

(sorry for the weird quote strings, I’m writing this on mobile…)

**Comment by Shatur** [comment]
The example looks valid at first look. Can't say what's wrong here.

**Comment by pianocomposer321** [comment]
Ope, figured it out, `g:asyncrun_silent` was set to 1.

**Comment by pianocomposer321** [comment]
Ok, so further testing has revealed that the autocommands do not fire when `-mode` is set to `term`. So unfortunately I can’t copy implementation from asyncrun. 😞

**Comment by Shatur** [comment]
That's bad, I can't migrate to yabs without this feature :(

**Comment by pianocomposer321** [comment]
Yeah, I’ll figure something out. It just can’t be the same as asyncrun.

**Comment by pianocomposer321** [comment]
I think what I’ll end up doing is something like what I said above. The config would look something like the following, for the same result as chaining commands before:

```lua
require(“yabs”):setup {
    languages = {
        cpp = {
            tasks = {
                build = {
                    command = “make build”,
                    output = “quickfix”
                },
                run = {
                    command = “./program”
                    output = “terminal”
                }
                build_and_run = {
                    command = function()
                        require(“yabs”).run_task(“build”)
                        -- logic to figure out when to run the run task…
                        require(“yabs”).run_task(“run”)
                    end,
                    type = “lua”
                }
            }
        }
    }
}
```

**Comment by Shatur** [comment]
> -- logic to figure out when to run the run task…

But how to figure the logic when to run the next task?

**Comment by pianocomposer321** [comment]
That’s the part I’m still trying to figure out :wink:.

I’ll probably have to start learning a bit more about how coroutines work in lua. That should be fun…

**Comment by pianocomposer321** [comment]
Ok, so here’s what I’m thinking…

I could add an `args` or `opts` table as a standard argument for all default `output` types, with `on_exit` or similar being a possible key. The value of this `on_exit` key would be a function that is called when the job finishes. This key would not be supported for `consolation` or `terminal` outputs, and users could decide for themselves whether they want to support it for their own output types.

```lua
tasks = {
    -- …
    build_and_run = {
        command = function()
            require(“yabs”):run_task(“build”, {on_exit = function() require(“yabs”):run_task(“run”)})
    }
    -- …
}
```

If I figure out coroutines, that api could probably be cleaned up quite a bit, maybe something more like

```lua
local build_task = require(“yabs”):run_task(“build”)
build_task:run_next(“run”)
```

**Comment by Shatur** [comment]
Sounds nice!

**Comment by Shatur** [comment]
Looks great to me!
Do you have any other plans for a `feature` branch?

**Comment by pianocomposer321** [comment]
I've decided that I'm going to keep the task "chaining" api as is for now. I might change it later, but for now it's not a priority. Unless @Shatur you have a specific suggestion for how you'd like it to look. If you do, I will probably implement it sooner rather than later, to prevent breaking changes after merging into master.
